### PR TITLE
switch to boto3 and add support for application ELBs with target groups.

### PR DESCRIPTION
**Pull request by jonhadfield**
ISSUE TYPE

Feature Pull Request
COMPONENT NAME

ec2_asg

ANSIBLE VERSION

ansible 2.2.0.0 (detached HEAD cdec853e37) last updated 2016/11/01 12:32:01 (GMT +100)
  lib/ansible/modules/core: (detached HEAD fe9c56a003) last updated 2016/11/01 12:32:05 (GMT +100)
  lib/ansible/modules/extras: (detached HEAD f564e77a08) last updated 2016/11/01 12:32:05 (GMT +100)
  config file =
  configured module search path = Default w/o overrides
SUMMARY

Upgrade module to use boto3 in order to support new functionality.
Added support for Application ELBs via Target Groups
Compliments the following PRs in that it enables creation of an ASG with ansible generated application load balancers and target groups:
elb_target_group: ansible/ansible-modules-extras#2947
elb_application_lb: ansible/ansible-modules-extras#2946

Note:

Only significant logic change is to support the integration with target groups/application lbs, so any existing issues remain
Only minimal testing performed

### CODE REVIEW

**Comment by jonhadfield** [comment]
This is a replacement for https://github.com/ansible/ansible-modules-core/pull/5746 after failed attempts at moving the original using PR mover.

File: lib/ansible/modules/cloud/amazon/ec2_asg.py
Comment (ryansb) [inline lib/ansible/modules/cloud/amazon/ec2_asg.py]: Needs error handling here for if no launch configs match the search, since later the `[0]` element is used without bounds-checking.

File: lib/ansible/modules/cloud/amazon/ec2_asg.py
Comment (ryansb) [inline lib/ansible/modules/cloud/amazon/ec2_asg.py]: This also needs bounds checking and (maybe) retries if the ASG isn't yet available.

File: lib/ansible/modules/cloud/amazon/ec2_asg.py
Comment (ryansb) [inline lib/ansible/modules/cloud/amazon/ec2_asg.py]: It looks like this call and the above (line 570) call are filling 2 variables with the same information. Is there a reason two names are used?

File: lib/ansible/modules/cloud/amazon/ec2_asg.py
Comment (ryansb) [inline lib/ansible/modules/cloud/amazon/ec2_asg.py]: This is a bit confusing since it's assigning to the same name name as comes in. I think this section could be cleared up with the use of `set`. So something along the lines of:

```
for dead_tag in set(have_tag_keyvals).difference(want_tag_keyvals):
    dead_tags.append(..... and so on ...)
```

File: lib/ansible/modules/cloud/amazon/ec2_asg.py
Comment (ryansb) [inline lib/ansible/modules/cloud/amazon/ec2_asg.py]: Please remove these commented out fail_json calls.

File: lib/ansible/modules/cloud/amazon/ec2_asg.py
Comment (ryansb) [inline lib/ansible/modules/cloud/amazon/ec2_asg.py]: Yay sets!

File: lib/ansible/modules/cloud/amazon/ec2_asg.py
Comment (ryansb) [inline lib/ansible/modules/cloud/amazon/ec2_asg.py]: :tada: boto3!

**Review by ryansb — DISMISSED** [review]

File: lib/ansible/modules/cloud/amazon/ec2_asg.py
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/ec2_asg.py]: Not sure I understand this change

File: lib/ansible/modules/cloud/amazon/ec2_asg.py
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/ec2_asg.py]: should be HAS_BOTO3, which you can import from module_utils.ec2

File: lib/ansible/modules/cloud/amazon/ec2_asg.py
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/ec2_asg.py:526]: All AWS calls should be wrapped in a `try`/`except` block

File: lib/ansible/modules/cloud/amazon/ec2_asg.py
Comment (s-hertel) [reply lib/ansible/modules/cloud/amazon/ec2_asg.py]: I wondered about this too. I think he chose this because iterating over the key/values in a dict has changed in python3 and module_utils.six checks for the python version so the behavior is consistent. It seems to behave appropriately. https://github.com/ansible/ansible/blob/devel/lib/ansible/module_utils/six.py#L367
Do you think this might end up being a problem?

File: lib/ansible/modules/cloud/amazon/ec2_asg.py
Comment (s-hertel) [reply lib/ansible/modules/cloud/amazon/ec2_asg.py]: Nice, will fix.

File: lib/ansible/modules/cloud/amazon/ec2_asg.py
Comment (jonhadfield) [reply lib/ansible/modules/cloud/amazon/ec2_asg.py]: @s-hertel Exactly that. Thought it best to make use of six for future compatibility with python3.

Comment (ryansb) [reply lib/ansible/modules/cloud/amazon/ec2_asg.py]: In Ansible we typically just use `.items()` because as a loop iterator there's no (usage) difference between the Py2 list and Py3 iterable behavior, and if your dictionary is less than 1000 items or so, the iterable behaviour is actually *slower* in Python 2 (trading time for space, naturally). Can you please leave it as `items` instead?

File: lib/ansible/modules/cloud/amazon/ec2_asg.py
Comment (ryansb) [reply lib/ansible/modules/cloud/amazon/ec2_asg.py:526]: I'm fine with this being a followup PR. Error handling needs to be added, but that'd increase the diffstat of this PR pretty significantly, and since the handling wasn't there before I wouldn't block the merge over it.

Comment (ryansb) [reply lib/ansible/modules/cloud/amazon/ec2_asg.py:526]: Strictly, they should also be using AWSRetry as well, but that can be left to a followup PR.

Comment (jonhadfield) [reply lib/ansible/modules/cloud/amazon/ec2_asg.py:526]: I'm calling this 16 times in total:
`connection.describe_auto_scaling_groups(AutoScalingGroupNames=[group_name])['AutoScalingGroups'][0]`
Will move it out to a function that includes a try/except block and then take a look at the remaining boto3 calls.

Comment (jonhadfield) [reply lib/ansible/modules/cloud/amazon/ec2_asg.py:526]: @ryansb Just seen your message re: diffstat. Will hold off on refactoring.

**Comment by jonhadfield** [comment]
Anything you're waiting on me for?

**Comment by s-hertel** [comment]
There are merge conflicts now to rebase, but not that I'm aware of. I think feature freeze was to blame. This still looks good to me.

**Comment by vanga** [comment]
There are some bugs in `delete_autoscaling_group` function. 
I have fixed and created a pull request for @jonhadfield branch.
https://github.com/jonhadfield/ansible/pull/1

File: lib/ansible/modules/cloud/amazon/ec2_asg.py
Comment (ryansb) [inline lib/ansible/modules/cloud/amazon/ec2_asg.py]: This isn't used anywhere

File: lib/ansible/modules/cloud/amazon/ec2_asg.py
Comment (ryansb) [inline lib/ansible/modules/cloud/amazon/ec2_asg.py:327]: What does this method name mean? No need to do anything, I'm just curious.

Comment (ryansb) [reply lib/ansible/modules/cloud/amazon/ec2_asg.py:327]: Oh, it's supposed to be "dereg" as in "deregister"

**Review by ryansb — COMMENTED** [review]
Looks good - I'll have to do some testing, but all feedback is there as a note to myself on things to fix later.

**Comment by tomislacker** [comment]
@jonhadfield I've also added metrics collection support in jonhadfield/ansible#2
EDIT: There's also a bug fix in there.

**Comment by s-hertel** [comment]
@tomislacker
Thanks for this! Can you please move your metrics collection support to a new PR since it's adding a new feature? I believe I've fixed the bugs but you might want to check me on that. You could keep the Fix notification setup commit though, and I'll add that. Good catch. :)

File: lib/ansible/modules/cloud/amazon/ec2_asg.py
Comment (s-hertel) [reply lib/ansible/modules/cloud/amazon/ec2_asg.py]: Fixed this on the line below.
describe_launch_configurations() is returning something like 
```
{'LaunchConfigurations': [], 'ResponseMetadata': {'RetryAttempts': 0, 'HTTPStatusCode': 200, 'RequestId': '57dccc7c-2080-11e7-8fbd-adf472434545', 'HTTPHeaders': {'x-amzn-requestid': '57dccc7c-2080-11e7-8fbd-adf472434545', 'date': 'Thu, 13 Apr 2017 19:35:25 GMT', 'content-length': '350', 'content-type': 'text/xml'}}}"
```
So it always had longer than a length of 0 and the below check had no effect. Added a check for the same thing on lines 719-720 too.

**Comment by jonhadfield** [comment]
Just re-tested after all the changes and needed to add a fix https://github.com/ansible/ansible/pull/19667/commits/1aa044f5efe4f196f8e763b889992e01ade1eb46 as changes to an existing ASG weren't being reflected in 'changed' status.

@tomislacker I've tested your PR https://github.com/jonhadfield/ansible/pull/2 for adding metrics collection support (locally only, not merged) and looks good! Only initial comments are:
- metrics_collection in docs is missing the 'default' 
- assuming additional metrics cost, shouldn't this be disabled by default?

@s-hertel asked to move it to a separate PR, so let me know when done and I'll re-test.

**Comment by s-hertel** [comment]
shipit

I removed all traces of boto.exception.stuff since it's no longer going to work here. I also didn't replace the exception handling around boto3_conn since boto3 fails when we try to use the connection rather than when it is created (whereas boto would give a NoAuthHandlerFound error when unable to get the connection). I kept a check in main in case the profile provided doesn't exist at all or bad credentials are provided. We still need to wrap the rest of the AWS calls in a try/except block as discussed above in a follow up PR - I just fixed the existing ones. 
Since this is passing the ec2_asg integration tests this gets a shipit from me.

@ryansb @willthames see any blockers for this now?

**Review by s-hertel — APPROVED** [review]

**Review by ryansb — APPROVED** [review]

**Comment by ryansb** [comment]
shipit

**Comment by ryansb** [comment]
shipit

**Comment by tomislacker** [comment]
@jonhadfield I've updated my branch for jonhadfield/ansible#2 for your review. Sorry for the delays, been a little too busy lately. Once this PR is merged, I can open one up on this project instead.

**Comment by shawnsi** [comment]
I've uncovered a potential concern with this update while integration testing.  In short, there are some parameters that worked implicitly and no longer do.  For example, `launch_config_name` would be pulled from any pre-existing ASG when the module was used.  This breaks ~~plays~~ tasks like this:

```
      ec2_asg:
        name: "my-asg"
        region: "us-east-1"
        health_check_type: ELB
```

This type of behavior is something used in my scenarios to toggle notifications, health checking, etc. back on after deployment activities.  This is a strange edge case as the parameters are marked required in documentation but the module was functional without them anyhow.  I'm open to refactoring our usage but it will be painful.  I imagine other users could hit the same problems as well.

Here is the error I see related to this when testing the new module:

```
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: Invalid type for parameter LaunchConfigurationNames[0], value: None, type: <type 'NoneType'>, valid types: <type 'basestring'>fatal: [us-east-1-xxx]: FAILED! => {"changed": false, "failed": true, "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/ansible_padsBj/ansible_module_ec2_asg.py\", line 1302, in <module>\n    main()\n  File \"/tmp/ansible_padsBj/ansible_module_ec2_asg.py\", line 1291, in main\n    create_changed, asg_properties = create_autoscaling_group(connection, module)\n  File \"/tmp/ansible_padsBj/ansible_modlib.zip/ansible/module_utils/cloud.py\", line 103, in retry_func\nbotocore.exceptions.ParamValidationError: Parameter validation failed:\nInvalid type for parameter LaunchConfigurationNames[0], value: None, type: <type 'NoneType'>, valid types: <type 'basestring'>\n", "module_stdout": "", "msg": "MODULE FAILURE"}```

**Comment by s-hertel** [comment]
Hi @shawnsi. Wow, I didn't even know that was functionality we had. `launch_config_name` is specified as required. My local integration tests do not test that edge case. I'll see if that's able to be reproduced with boto3 and I'll make a fix (and correct the documentation) if so. Thanks for being another set of eyes on this.

@tomislacker No problem, thanks for doing that! Looks great. I will cherry pick fd68a1078a1841aad22f43ef3dfb0de3aa68fc6f and 9705587401878d4f9ba229ea2be739ce9d6bfbc2.

**Comment by s-hertel** [comment]
@shawnsi Can you please run your tests again? I think it should work now.

**Comment by tomislacker** [comment]
@s-hertel I was able to successfully test out the workflow that @shawnsi mentioned previously.

**Comment by s-hertel** [comment]
shipit

@ryansb Can you check this out again when you get the chance?

**Review by ryansb — APPROVED** [review]

**Comment by ryansb** [comment]
shipit

**Comment by avishefi** [comment]
Hi,
I noticed that ec2_asg doesn't deregister instances from ALB target groups.
Is this by design? Do ELB fail to deregister instances but ALBs do that out of the box and therefore this was not implemented in ec2_asg?

**Comment by willthames** [comment]
@avishefi - I'd think deregistering instances should be done through elb_target_groups - you disassociate the ASG from the TG.

If there's a use case that isn't working, let us know, preferably in a separate issue now that this  one is closed.

**Pull request merged**
